components/proxy/src/main/java/com/hotels/styx/StyxServer.java

```diff
@@ -131,34 +143,47 @@ private static boolean skipServerConfigValidation() {
         return "n".equals(validate) || "no".equals(validate);
     }
 
-    private final HttpServer proxyServer;
-    private final HttpServer adminServer;
+    private final ConfigStore configStore;
+    private final long startupTimeoutSeconds;
 
     private final ServiceManager serviceManager;
     private final Stopwatch stopwatch;
 
+    private final ServerReference proxyServer = new ServerReference("Proxy server");
+    private final ServerReference adminServer = new ServerReference("Admin server");
+
     public StyxServer(StyxServerComponents config) {
         this(config, null);
     }
 
     public StyxServer(StyxServerComponents components, Stopwatch stopwatch) {
+        this.startupTimeoutSeconds = components.environment().configuration()
+                .get("startup.timeoutSeconds", Integer.class)
+                .map(timeout -> (long) timeout)
+                .orElse(MAX_VALUE);
+
         this.stopwatch = stopwatch;
 
         registerCoreMetrics(components.environment().buildInfo(), components.environment().metricRegistry());
 
         Map<String, StyxService> servicesFromConfig = components.services();
 
-        ProxyServerSetUp proxyServerSetUp = new ProxyServerSetUp(new StyxPipelineFactory());
+        PluginStartupService pluginStartupService = new PluginStartupService(components);
+
+        this.configStore = components.environment().configStore();
 
-        components.plugins().forEach(plugin -> components.environment().configStore().set("plugins." + plugin.name(), plugin));
+        StyxService adminServerService = new ServerService("adminServer", () ->
+                this.adminServer.store(createAdminServer(components)));
```

Comment:
Why it is necessary to an `HttpServer` instance and wrap it into a `ServerReference` object?

Why not to create an `HttpServer` instance and pass it as is to `ServerService`? Like so:

    new ServerService(createAdminServer(components));

Reply:
The whole point of the `ServerService` is to run the creation/start-up process of the server. If you do it before construction, then the service does nothing, and we would not be able to start the servers asynchronously.
The `ServerReference` is because the `StyxServer` needs to receive the server after it has been created asynchronously.

Reply:
`createAdminServer()` just creates a server object. It doesn't start it. The service is still be able to start it by invoking `startAsync`.

Okay. I see. The `createProxyServer` blocks until the plugins are loaded. This is somewhat confusing:

`createProxyServer()` returns an `HttpServer` instance which has `startAsync` method. Typical usage is:

    HttpServer server = createProxyServer();
    server.startAsync().awaitRunning();

Where does it block? On both lines. Surprisingly. The first line blocks at loading the plugins, the 2nd line blocks at waiting for the server to start.

IMO it would be more sensible to wait for plugins in the service startup method, rather than in HttpServer creation.